pkg/updater/updater.go

```diff
@@ -0,0 +1,224 @@
+package updater
+
+import (
+	"fmt"
+	"github.com/go-git/go-billy/v5"
+	"github.com/go-git/go-billy/v5/memfs"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"github.com/rs/zerolog/log"
+	"gopkg.in/yaml.v2"
+	"io"
+	"os"
+	"time"
+)
+
+var (
+	sshKeyPath    = os.Getenv("SSH_KEY_PATH")
+	sshKeyPass    = os.Getenv("SSH_KEY_PASS")
+	sshCommitUser = os.Getenv("SSH_COMMIT_USER")
+	sshCommitMail = os.Getenv("SSH_COMMIT_MAIL")
+)
+
+type ArgoOverrideFile struct {
+	Helm struct {
+		Parameters []ArgoParameterOverride `yaml:"parameters"`
+	} `yaml:"helm"`
+}
+
+type ArgoParameterOverride struct {
+	Name        string `yaml:"name"`
+	Value       string `yaml:"value"`
+	ForceString bool   `yaml:"forceString"`
+}
+
+type GitRepo struct {
+	RepoURL    string
+	BranchName string
+	Path       string
+	fs         billy.Filesystem
+	localRepo  *git.Repository
+	sshAuth    *ssh.PublicKeys
+}
+
+func (repo *GitRepo) Clone() error {
+	var err error
+
+	repo.fs = memfs.New()
+
+	if repo.sshAuth, err = ssh.NewPublicKeysFromFile("git", sshKeyPath, sshKeyPass); err != nil {
+		return err
+	}
+
+	repo.localRepo, err = git.Clone(memory.NewStorage(), repo.fs, &git.CloneOptions{
+		URL:           repo.RepoURL,
+		ReferenceName: plumbing.ReferenceName("refs/heads/" + repo.BranchName),
+		SingleBranch:  true,
+		Auth:          repo.sshAuth,
+	})
+
+	return err
+}
+
+func (repo *GitRepo) UpdateApp(appName string, overrideContent *ArgoOverrideFile) error {
+	overrideFileName := fmt.Sprintf("%s/.argocd-source-%s.yaml", repo.Path, appName)
+	commitMsg := fmt.Sprintf("argo-watcher(%s): update image tag", appName)
+
+	overrideContent, err := repo.mergeOverrideFileContent(overrideFileName, overrideContent)
+	if err != nil {
+		return err
+	}
+
+	if err := repo.commit(overrideFileName, commitMsg, overrideContent); err != nil {
```

Comment:
When does the push happen?

Reply:
When the override file is either absent or its value has changed.